SA-258 add table diffs

cordeezy2 opened the pull request.

cordeezy2 committed:
Table Diffs

cordeezy2 committed:
fix for snapshot names

cordeezy2 commented:
@justinjeffery-ipf Thoughts on this?
any recommendations ?

cordeezy2 committed:
black fixes

ghost commented:
I would think this would just be a function in the Table class: https://github.com/community-fabric/python-ipfabric/blob/00fdaf3f113b0ecb8daab27867db95db9fa0d360/ipfabric/models.py#L11

This would be similar to how Intents work: https://github.com/community-fabric/python-ipfabric/blob/00fdaf3f113b0ecb8daab27867db95db9fa0d360/ipfabric/intent.py#L103

ghost commented:
Default to ignore certain columns:

- id
- _id
- rd (Routing Domain)
- stpDomain

User must be able to specify which columns they want to compare (and/or columns to ignore).  For instance Inventory > Devices ignore `uptime` and `memoryUtilization`.

Using deepdiff:
https://zepworks.com/deepdiff/current/exclude_paths.html#exclude-paths-label

ghost commented:
Investigate deepdiff optimization: https://zepworks.com/deepdiff/current/optimizations.html

Example benchmark: https://github.com/seperman/deepdiff-benchmark

Probably use Routing Table for testing?

cordeezy2 committed:
Squashed commit of the following:

commit 39a169a9d0c2631e82676a2fa30f64e91c437ebf
Merge: 1e0c208 ba7a829
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 10:52:35 2022 -0500

    Merge pull request #181 from community-fabric/ipf_v6.0.1

    Update macaddress

commit ba7a82983b8beebb52c1156bd014ca8764e04b6f
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Wed Nov 16 10:51:56 2022 -0500

    Update macaddress

commit b0df4c844c6aef7dd4df308ba607c12eeeadfc66
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Wed Nov 16 10:47:17 2022 -0500

    Update macaddress

commit 1e0c208c213c5811b998e69d2dc457b6d55e87a5
Merge: 66f5c02 2a9692d
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 10:08:44 2022 -0500

    Merge pull request #180 from community-fabric/ipf_v6.0

    Fix interfaces

commit 2a9692dbeee9a7432b56be7aca853d150b4f2415
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Wed Nov 16 10:04:27 2022 -0500

    Fix interfaces
    Update functional tests

commit 66f5c0245cee6cc7f5cd61c652e85914d18d37aa
Merge: b85e11b 5c643aa
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 09:38:51 2022 -0500

    Merge pull request #179 from community-fabric/ipf_v6.0

    Poetry Update

commit 5c643aabcf3b8749fbcfde296302552ae1336888
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Mon Nov 7 09:07:28 2022 -0500

    Poetry Update

commit b85e11b985ed6a73143f5a0e8af91e510cf0e1e4
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 09:34:42 2022 -0500

    Update CHANGELOG.md

commit 77d712412d4be874b99afbfc3d2947882bcb3924
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 09:31:42 2022 -0500

    Update CHANGELOG.md

commit 91a629ad838ac97d30d17f9db5c0f94a5795037d
Merge: 00fdaf3 992adaa
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Thu Nov 10 15:01:50 2022 -0500

    Merge pull request #177 from community-fabric/NIM-8924

    RBAC Examples

commit 992adaaee94455a7755def9e5180913bff1a2043
Author: Alex Gittings <agitting96@gmail.com>
Date:   Thu Nov 10 10:35:41 2022 +0000

    update README.md for RBAC example

commit 6754c40e654e578e134e98b57efc6ff9bb98235b
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 17:52:05 2022 +0000

    add new policies and role

commit 88229772bbac31a1f80bf403b8f107be6d2a1ccf
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 11:58:09 2022 +0000

    fix strings and reorder python imports

commit 2534467a9b6fbdac07ecaebf47737589082ef257
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 11:55:21 2022 +0000

    fix some bits

commit 5c5fed144e319310e95cc3cfa597c11fb64114f0
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 11:35:41 2022 +0000

    Initial Draft

cordeezy2 committed:
Squashed commit of the following:

commit 074f7a4b87cd2a9d59d7f20d210c1fa0d56b39d5
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Thu Nov 17 10:37:18 2022 -0500

    Update CHANGELOG.md

commit 2601be4f4f97b8892cb24dc16a03fa37da742956
Merge: 39a169a b1faebc
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Thu Nov 17 10:36:31 2022 -0500

    Merge pull request #182 from community-fabric/ipf_v6.0

    Fix for Attributes

commit b1faebcc19a88246f2ea1f4c8c54f6cfb3b5b1bc
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Thu Nov 17 10:33:45 2022 -0500

    Fix for Attributes

commit 39a169a9d0c2631e82676a2fa30f64e91c437ebf
Merge: 1e0c208 ba7a829
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 10:52:35 2022 -0500

    Merge pull request #181 from community-fabric/ipf_v6.0.1

    Update macaddress

commit ba7a82983b8beebb52c1156bd014ca8764e04b6f
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Wed Nov 16 10:51:56 2022 -0500

    Update macaddress

commit b0df4c844c6aef7dd4df308ba607c12eeeadfc66
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Wed Nov 16 10:47:17 2022 -0500

    Update macaddress

commit 1e0c208c213c5811b998e69d2dc457b6d55e87a5
Merge: 66f5c02 2a9692d
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 10:08:44 2022 -0500

    Merge pull request #180 from community-fabric/ipf_v6.0

    Fix interfaces

commit 2a9692dbeee9a7432b56be7aca853d150b4f2415
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Wed Nov 16 10:04:27 2022 -0500

    Fix interfaces
    Update functional tests

commit 66f5c0245cee6cc7f5cd61c652e85914d18d37aa
Merge: b85e11b 5c643aa
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 09:38:51 2022 -0500

    Merge pull request #179 from community-fabric/ipf_v6.0

    Poetry Update

commit 5c643aabcf3b8749fbcfde296302552ae1336888
Author: Justin Jeffery <justin.jeffery@ipfabric.io>
Date:   Mon Nov 7 09:07:28 2022 -0500

    Poetry Update

commit b85e11b985ed6a73143f5a0e8af91e510cf0e1e4
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 09:34:42 2022 -0500

    Update CHANGELOG.md

commit 77d712412d4be874b99afbfc3d2947882bcb3924
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Wed Nov 16 09:31:42 2022 -0500

    Update CHANGELOG.md

commit 91a629ad838ac97d30d17f9db5c0f94a5795037d
Merge: 00fdaf3 992adaa
Author: Justin Jeffery <92743314+justinjeffery-ipf@users.noreply.github.com>
Date:   Thu Nov 10 15:01:50 2022 -0500

    Merge pull request #177 from community-fabric/NIM-8924

    RBAC Examples

commit 992adaaee94455a7755def9e5180913bff1a2043
Author: Alex Gittings <agitting96@gmail.com>
Date:   Thu Nov 10 10:35:41 2022 +0000

    update README.md for RBAC example

commit 6754c40e654e578e134e98b57efc6ff9bb98235b
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 17:52:05 2022 +0000

    add new policies and role

commit 88229772bbac31a1f80bf403b8f107be6d2a1ccf
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 11:58:09 2022 +0000

    fix strings and reorder python imports

commit 2534467a9b6fbdac07ecaebf47737589082ef257
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 11:55:21 2022 +0000

    fix some bits

commit 5c5fed144e319310e95cc3cfa597c11fb64114f0
Author: Alex Gittings <agitting96@gmail.com>
Date:   Wed Nov 9 11:35:41 2022 +0000

    Initial Draft

cordeezy2 committed:
compare method for tables

looping over each returned value, turn that value into a string, hash it.
compare the hashes to determine what the diffs are, return the diffed data.

cordeezy2 commented:
@justinjeffery-ipf 
heres the hashing idea i discussed this morning, let me know what you think. 
I like this method, as it doesnt introduce any additional package requirements and still provides the basic functionality requested.

TODO:
add some benchmarks for performance of diffs

Commit:
Merge pull request #185 from community-fabric/develop

Pull develop changes

ghost commented:
@cordeezy2 merge develop into branch so I could check changes

Commit:
Move get_snapshot_id to client so it can be reused.
Removed some functions and simplified table_diff.py

Commit:
Fix for import

Commit:
Cleaned up _check_tech_table_attr

justinjeffery-ipf commented on ipfabric/tools/table_diff.py:
```diff
@@ -0,0 +1,201 @@
+            logger.info(f"valid tables: \n {valid_tables}")
+            logger.info(f"{HELP_STR}")
+
+    def _check_inv_table_attr(self, inv_table: str):
```
Is this still needed?  Don't see it referenced anywhere

cordeezy2 replied:
I removed this file, sorry for the confusion, my bad :pensive:

cordeezy2 committed:
Merge pull request #186 from community-fabric/SA-258-add-table-diffs-justin

Simplified table_diff.py

cordeezy2 committed:
removing tablediffer

cordeezy2 committed:
columns ignore fix

cordeezy2 committed:
update for cols_ignore

cordeezy2 commented on ipfabric/models.py:
```diff
@@ -104,6 +105,94 @@ def count(
+    def _compare_determine_columns(self, table_columns, columns, columns_ignore): #NOSONAR
+        cols_for_return = []
+        # user does not pass any columns
+        if columns is None and columns_ignore is None:
```
2250f9f corrected in latest commits

cordeezy2 commented on ipfabric/models.py:
```diff
@@ -104,6 +105,94 @@ def count(
+        table_cols = self.client._get_columns(self.endpoint)
+
+        # determine which columns to use in query
+        if columns or columns_ignore:
```
corrected in latest commits 2250f9f

cordeezy2 commented on ipfabric/tools/table_diff.py:
```diff
@@ -0,0 +1,201 @@
+    from ipfabric.snapshot_models import Snapshot
+import logging
+from typing import Any
+import jsondiff as jd
```
removed file

Commit:
Updated models.py

Commit:
Updated models.py

cordeezy2 committed:
removing nosonar comment

cordeezy2 committed:
Merge pull request #187 from community-fabric/SA-258-add-table-diffs-justin

determine columns updated in models.py

cordeezy2 committed:
unit test for compare

cordeezy2 committed:
typing and doc strings

cordeezy2 committed:
3.8 fix for typing

cordeezy2 committed:
Merge branch 'develop' into SA-258-add-table-diffs

cordeezy2 commented:
@justinjeffery-ipf should be ready for merge, let me know if theres anything else missing

cordeezy2 committed:
typing fix

cordeezy2 commented:
all Test are passing in my local env, odd

cordeezy2 commented:
i was testing on 3.8.13, CI is running on 3.8.14

cordeezy2 committed:
removing typing

typing is breaking python3.8 CI

cordeezy2 committed:
.env support for locally testing
